### PR TITLE
spinel-cli: Add NCP build to TravisCI

### PR DESCRIPTION
**Pull request by turon**
This PR adds a command line tool, named spinel-cli, that converts ot-cli shell commands (as documented in https://github.com/openthread/openthread/blob/master/src/cli/README.md) to spinel NCP commands (as documented in https://github.com/openthread/openthread/blob/master/src/ncp/PROTOCOL.md).  

Also added is a patch to node.py that provides a way to run the existing continuous integration certification test suite against an NCP build rather than a CLI build (by prefacing the command to start the test with the environment variable setting `NODE_TYPE=ncp-sim`):

```
cd tests/scripts/thread-cert
NODE_TYPE=ncp-sim top_builddir=../../.. python Cert_5_1_02_ChildAddressTimeout.py 
```

The power of this tool is three fold: 
1) As a path to add testing of the NCP in simulation to continuous integration
2) As a path to automated testing of testbeds running NCP firmware on hardware
3) As a simple debugging tool for NCP builds of OpenThread

```
openthread$ cd tools/spinel-cli/
spinel-cli$ ./spinel-cli.py
Opening pipe to ../../examples/apps/ncp/ot-ncp 1
spinel-cli > help

Available commands (type help <name> for more information):
============================================================
channel            extaddr       mode              route                 
child              extpanid      netdataregister   router                
childtimeout       h             networkidtimeout  routerupgradethreshold
clear              help          networkname       scan                  
contextreusedelay  history       panid             state                 
counter            ifconfig      ping              thread                
debug              ipaddr        prefix            v                     
debug-term         keysequence   q                 version               
eidcache           leaderdata    quit              whitelist             
enabled            leaderweight  releaserouterid 
exit               masterkey     rloc16          

spinel-cli > version
OPENTHREAD/gd4d4e9d-dirty; Aug 11 2016 14:40:44
Done
spinel-cli > thread start
Done
spinel-cli > state
leader
Done
spinel-cli > 
```

The tool provides three ways to connect to an NCP image:
   1) pipe to stdin/stdout of a command line tool (default): -p ./ot-ncp
   2) serial connection to a hardware device: -u /dev/ttyUSB0
   3) socket connection to the port of a helper tool such as ncp-spi-driver: -s 1234 

```
spinel-cli$ ./spinel-cli.py --help
Options:
  -h, --help            show this help message and exit
  -u UART, --uart=UART  
  -p PIPE, --pipe=PIPE  
  -s SOCKET, --socket=SOCKET
  -n NODEID, --nodeid=NODEID
  -q, --quiet           
  -v, --verbose         
```

This is currently a work-in-progress, but there is enough up and running to let folks take an early look.

### CODE REVIEW

**Comment by mszczodrak** [comment]
LGTM :+1:

**Comment by turon** [comment]
This is ready to go!

File: .travis/before_install.sh [hunk 1]
```diff
@@ -55,6 +55,13 @@ cd /tmp || die
     [ $BUILD_TARGET != posix-32-bit ] || {
         sudo apt-get install g++-multilib || die
     }
+
+    [ $NODE_TYPE != ncp-sim ] || {
+	sudo easy_install pip || die
```
Comment (jwhui) [inline .travis/before_install.sh]: fix whitespace

Comment (jwhui) [inline .travis/before_install.sh]: fix whitespace

File: src/ncp/PROTOCOL.md [hunk 2]
```diff
@@ -1624,6 +1624,15 @@ Note that some implementations may not support `CMD_GET_VALUE`
 routerids, but may support `CMD_REMOVE_VALUE` when the node is
 a leader.
 
+#### D.4.22. PROP 5382: `PROP_THREAD_RLOC16_DEBUG_PASSTHRU`
```
Comment (jwhui) [inline src/ncp/PROTOCOL.md]: Please separate this property addition into a separate commit.

File: tests/scripts/thread-cert/Cert_5_1_02_ChildAddressTimeout.py [hunk 3]
```diff
@@ -99,7 +99,7 @@ def test(self):
             if addr[0:4] != 'fe80':
                 try:
                     self.nodes[LEADER].ping(addr)
-                    self.assertFalse()
+                    self.assertFalse(False)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_1_02_ChildAddressTimeout.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/Cert_5_1_02_ChildAddressTimeout.py [hunk 4]
```diff
@@ -109,7 +109,7 @@ def test(self):
...
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_1_02_ChildAddressTimeout.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/Cert_5_3_03_AddressQuery.py [hunk 5]
```diff
@@ -138,7 +138,7 @@ def test(self):
             if addr[0:4] != 'fe80':
                 try:
                     self.nodes[ED2].ping(addr)
-                    self.assertFalse()
+                    self.assertFalse(False)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_3_03_AddressQuery.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/Cert_5_3_03_AddressQuery.py [hunk 6]
```diff
@@ -149,7 +149,7 @@ def test(self):
             if addr[0:4] != 'fe80':
                 try:
                     self.nodes[BR].ping(addr)        
-                    self.assertFalse()
+                    self.assertFalse(False)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_3_03_AddressQuery.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py [hunk 7]
```diff
@@ -117,7 +117,13 @@ def test(self):
         self.nodes[ED2].add_ipaddr('2001::1')
         time.sleep(3)
 
-        self.nodes[ED3].ping('2001::1')
+
+        # Final ping doesn't need a response;
+        # used to trigger an ADDR_ERR.ntf
+        try:
+            self.nodes[ED3].ping('2001::1')
+        except pexpect.TIMEOUT:
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: I think the final ping should still generate an Echo Reply since the first one to respond will still fill the address cache.

File: tests/scripts/thread-cert/Cert_5_3_10_AddressQuery.py [hunk 8]
```diff
@@ -128,7 +128,7 @@ def test(self):
             if addr[0:4] != 'fe80':
                 try:
                     self.nodes[SED2].ping(addr)
-                    self.assertFalse()
+                    self.assertFalse(False)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_3_10_AddressQuery.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/Cert_5_3_10_AddressQuery.py [hunk 9]
```diff
@@ -140,7 +140,7 @@ def test(self):
             if addr[0:4] != 'fe80':
                 try:
                     self.nodes[BR].ping(addr)
-                    self.assertFalse()
+                    self.assertFalse(False)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_3_10_AddressQuery.py]: `False` is not needed here.  Empty parameters default to `False` condition.

File: tests/scripts/thread-cert/node.py [hunk 10]
```diff
@@ -68,15 +70,34 @@ def __init_sim(self, nodeid):
         # Add delay to ensure that the process is ready to receive commands.
         time.sleep(0.1)
 
+
+    def __init_ncp_sim(self, nodeid):
+        """ Initialize an NCP simulation node. """
+        if "top_builddir" in os.environ.keys():
+            srcdir = os.environ['top_builddir']
+            dir_path = os.path.dirname(os.path.realpath(__file__))
```
Comment (jwhui) [inline tests/scripts/thread-cert/node.py]: Use `top_srcdir` instead?

**Comment by jwhui** [comment]
Please add a `README.md` that describes this tool and how to use it.

File: src/ncp/PROTOCOL.md (hunk 2 above)
Comment (turon) [inline src/ncp/PROTOCOL.md]: Ok, done in this PR: https://github.com/openthread/openthread/pull/493

File: tools/spinel-cli/README.md [hunk 11]
```diff
@@ -0,0 +1,1327 @@
...
+### Run entire NCP thread-cert suite
+
+```
+# From top-level of openthread tree
+make distclean
+./bootstrap
+NODE_TYPE=ncp-sim BUILD_TARGET=posix-distcheck DISTCHECK_CONFIGURE_FLAGS="--with-examples=posix --enable-cli --enable-ncp=uart --with-tests=all" make -f examples/Makefile-posix distcheck BuildJobs=10 VERBOSE=1
+```
+
+## Command Reference
```
Comment (jwhui) [inline tools/spinel-cli/README.md]: The Command Reference should just refer to [src/cli/README.md](https://github.com/openthread/openthread/blob/master/src/cli/README.md)

File: tools/spinel-cli/spinel-cli.py [hunk 12]
```diff
@@ -0,0 +1,2953 @@
+#!/usr/bin/python
```
Comment (jwhui) [inline tools/spinel-cli/spinel-cli.py]: Missing file header.

File: tools/spinel-cli/README.md [hunk 13]
```diff
@@ -0,0 +1,1327 @@
+# Spinel CLI Reference
```
Comment (jwhui) [inline tools/spinel-cli/README.md]: Please spellcheck this doc.

File: tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py (hunk 7 above)
Comment (turon) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: In the NCP configuration, this assumption fails: no response is raised to the host and the ping timeout is exceed.  It seems the intent of this test is to verify that ADDR_ERR.ntf is sent and not that a ping response comes in during a duplicate address situation.

We can investigate why NCP and raw CLI behavior is different here, but I don't think it is worth blocking addition of continuous integration of Spinel NCP configurations for a check that isn't actually part of the certification test plan.

Comment (turon) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: Here is the debug output for Cert_5_3_07 after setting up the duplicate address situation.  This trace is from when the ping is sent to the initial timeout.  It looks like after sending the ping, the host gets back "no route", no response, and then the timeout.

```
### ===> Node 6 issues CLI command to HOST driver
6: ping 2001::1

### ===> Some asynchronous inbound traffic before command is issued at Spinel level
spinel-cli > RX Pay: (99) ['80', '06', '73', '5C', '00', '60', '00', '00', '00', '00', '34', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'F0', 'D9', '2A', '82', 'C8', 'D8', 'FE', '43', 'FF', '02', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '01', '4D', '4C', '4D', '4C', '00', '34', 'F2', '78', '00', '15', '10', '00', '00', '00', '00', '00', '00', '00', '01', '16', '47', 'FC', '46', '5B', '58', 'F7', '15', 'D5', '7C', '36', '52', 'EA', 'CD', '23', 'CE', '96', 'DD', '3D', '24', 'BC', 'DC', '97', 'DF', 'E6', '56', '03', 'C9', '72', '29', 'A5', '11', 'B9', 'F5', 'F3'] 
RX Pay: (87) ['80', '06', '73', '50', '00', '60', '00', '00', '00', '00', '28', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'F0', 'D9', '2A', '82', 'C8', 'D8', 'FE', '43', 'FF', '02', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '01', '4D', '4C', '4D', '4C', '00', '28', '60', '02', '00', '15', '11', '00', '00', '00', '00', '00', '00', '00', '01', '2B', 'D8', '6E', 'B1', '65', '1F', '36', '1E', '11', '14', '98', '03', '27', 'C4', '70', 'BB', '9F', '94', '62', '1F', '78', '23', 'D0'] 
ping 2001::1
TX Pay: (3) ['81', '02', '61'] 
RX Pay: (5) ['80', '06', '5A', '2F', '63'] 
PROP_VALUE_IS: THREAD_STABLE_NETWORK_DATA = None
RX Pay: (105) ['80', '06', '73', '62', '00', '60', '00', '00', '00', '00', '3A', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'F0', 'D9', '2A', '82', 'C8', 'D8', 'FE', '43', 'FE', '80', '00', '00', '00', '00', '00', '00', 'A4', '1A', 'FE', '0F', 'B1', '10', 'F8', '96', '4D', '4C', '4D', '4C', '00', '3A', 'F9', 'B6', '00', '15', '12', '00', '00', '00', '00', '00', '00', '00', '01', 'A7', '5D', 'A1', '19', '11', '80', '4C', '38', '0D', 'D8', '00', '95', 'EA', '04', 'F5', '7C', '9A', '83', '1A', '43', '90', '59', '79', 'BF', 'EE', 'EB', '3D', '6F', 'DE', '62', 'A3', 'F1', '8D', '5B', 'E3', '41', '23', 'E1', '33', '92', '16'] 
RX Pay: (21) ['81', '06', '61', 'FD', 'DE', 'AD', '00', 'BE', 'EF', '00', '00', '23', '15', 'C1', '82', '07', '32', '2E', '9B', '2C', '1F'] 

### ===> Actual send of ping6 2001::1 over spinel to NCP
TX Pay: (55) ['81', '03', '72', '30', '00', '60', '00', '00', '00', '00', '08', '3A', '40', 'FD', 'DE', 'AD', '00', 'BE', 'EF', '00', '00', '23', '15', 'C1', '82', '07', '32', '2E', '9B', '20', '01', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '01', '80', '00', 'DB', '86', '00', '00', '00', '00', '00', '00'] 

### ===> Spinel returns a Thread stack error
RX Pay: (8) ['81', '06', '00', '84', '80', '01', '56', '3F'] 
PROP_VALUE_IS: LAST_STATUS = 16388
kThreadError_NoRoute
E1: thread stop
```

Comment (turon) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: @darconeous @jwhui 
Given that this test is one of the only to excercise manual address configuration, and the NoRoute error idicates some issue with the on-mesh prefix configuration, I looked through the log again. It looks like earlier in the test, the insert prefix command is failing:
```
prefix add 2001::/64 paros med
### ==> 0x5A = SPINEL_PROP_THREAD_ON_MESH_NETS
### ==> add prefix 2001::/64,  stable=1, flags=0x73
TX Pay: (22) ['81', '04', '5A', '20', '01', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '40', '01', '73'] 
RX Pay: (91) ['80', '06', '73', '54', '00', '60', '00', '00', '00', '00', '2C', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'B4', '40', 'D4', '8C', 'E9', '38', 'F9', '52', 'FF', '02', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '00', '02', '4D', '4C', '4D', '4C', '00', '2C', 'B6', '50', '00', '15', '00', '00', '00', '00', '00', '00', '00', '00', '01', '44', '76', 'D6', 'FD', '36', '6F', 'EA', 'EA', 'CC', 'BC', 'F5', '02', 'A3', '01', 'D9', '97', '20', 'D6', 'CB', 'E0', '92', 'B1', 'C4', 'B3', 'FA', 'FE', '32'] 
RX Pay: (122) ['80', '06', '73', '73', '00', '60', '00', '00', '00', '00', '4B', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'B4', '40', 'D4', '8C', 'E9', '38', 'F9', '52', 'FE', '80', '00', '00', '00', '00', '00', '00', 'D0', '8D', '7F', '87', '58', '88', 'FC', 'CB', '4D', '4C', '4D', '4C', '00', '4B', '27', '06', '00', '15', '01', '00', '00', '00', '00', '00', '00', '00', '01', 'EF', 'F7', '65', '51', 'E5', 'D4', '88', '89', '96', 'D1', '76', '1F', '16', '7A', '0E', '36', '01', '75', 'A7', '36', 'CB', '32', 'B9', 'EE', '94', 'B3', 'DD', '8D', 'B0', '1D', '6A', '22', 'E2', '76', '14', '59', 'B9', '34', 'B3', '80', '9C', '14', 'BB', 'FB', 'B7', '9C', '08', '7B', '58', '74', '36', 'D8', 'DC', '18', '49', '83', '09', '52'] 
RX Pay: (105) ['80', '06', '73', '62', '00', '60', '00', '00', '00', '00', '3A', '11', 'FF', 'FE', '80', '00', '00', '00', '00', '00', '00', 'B4', '40', 'D4', '8C', 'E9', '38', 'F9', '52', 'FE', '80', '00', '00', '00', '00', '00', '00', 'D0', '8D', '7F', '87', '58', '88', 'FC', 'CB', '4D', '4C', '4D', '4C', '00', '3A', '55', '9A', '00', '15', '02', '00', '00', '00', '00', '00', '00', '00', '01', '00', 'C8', '07', 'CD', 'C3', '87', '62', '71', '5C', '83', '36', '51', '75', '49', 'F4', '38', '17', '07', '33', '54', '3E', '14', '69', 'F8', '26', 'EE', 'E4', '08', '85', '42', 'A0', '2E', '3A', '2E', '18', 'F9', 'FE', '20', 'CF', '5E', '5D'] 
RX Pay: (6) ['81', '06', '00', '04', 'F6', '5D'] 
PROP_VALUE_IS: LAST_STATUS = 4
STATUS_INVALID_STATE : The given operation is invalid for the current state of the device.
Done
```

Comment (darconeous) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: It seems that doing a set-true/set-false of `PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE` after all the changes isn't sufficient, due to the following code in both `InsertPropertyHandler_THREAD_LOCAL_ROUTES()` and `InsertPropertyHandler_THREAD_ON_MESH_NETS()`, added in #273 (ping @abtink):

```
    VerifyOrExit(
        mAllowLocalNetworkDataChange == true,
        errorCode = SendLastStatus(header, SPINEL_STATUS_INVALID_STATE)
    );
```

This check is present to force the host implementation to use it in a way that is consistent with how we intended it to be used (Set this property to `true`, make your changes, then set it to `false`). I see three options to proceed forward:

1. Remove these checks from those functions in `ncp_base.cpp` and continue what you are doing.
2. Make the NCP automatically set `PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE` to `true` whenever a net data related property is modified.
3. Make sure your calls to add prefixes and local routes set `PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE` to true before setting the prefix. It is totally OK to set it to true multiple times.

I'm inclined to think option 3 would be the best since these checks force the host to interact with the NCP in the prescribed way—which should hopefully lead to better host implementations.

Comment (turon) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: Option (3) sounds good.  Thanks for the detailed next step plan.

Comment (turon) [inline tests/scripts/thread-cert/Cert_5_3_07_DuplicateAddress.py]: @darconeous, option (3) has been implemented, and now the prefix add call returns success.  The ping still fails with NoRoute however.  Do the `ipaddr add` commands need to set PROP_THREAD_ALLOW_LOCAL_NET_DATA_CHANGE=1 as well, or something similar?

**Comment by darconeous** [comment]
By the way, I've written a commit which adds an XFAIL for the expected test failure: 8db5c7f83bdaa9470813224351f921817c5a78f4 (Travis is running [here](https://travis-ci.org/darconeous/openthread/jobs/156950803))

It is in my [darconeous/openthread](https://github.com/darconeous/openthread/) repository in the branch [feature/test/ncp](https://github.com/darconeous/openthread/tree/feature/test/ncp).

**Comment by darconeous** [comment]
I [see a lot more failures](https://travis-ci.org/darconeous/openthread/jobs/156950803#L1471) than just `Cert_5_3_07_DuplicateAddress` after this change to `node.py` from 8db5c7f83bdaa9470813224351f921817c5a78f4:

```diff
diff --git a/tests/scripts/thread-cert/node.py b/tests/scripts/thread-cert/node.py
index b699125..d427626 100755
--- a/tests/scripts/thread-cert/node.py
+++ b/tests/scripts/thread-cert/node.py
@@ -98,8 +98,6 @@ class Node:
 
     def __del__(self):
         if self.pexpect.isalive():
-            if self.node_type == 'ncp-sim':
-                self.pexpect.sendcontrol('c');
             self.pexpect.terminate()
             self.pexpect.close(force=True)
 
```

**Comment by turon** [comment]
@darconeous Yes, spinel-cli needs to be careful to properly close its ot-ncp child process.  If those linger around, you get multiple nodes with the same id and it causes failures.

**Comment by jwhui** [comment]
I just added a bunch more addressing checks to the border routing tests in #512.  I'm curious to see if spinel-cli can pass those extra checks.

**Comment by darconeous** [comment]
turon: What happens if a packet happens to contain `0x03`? (CTRL-C)

**Comment by turon** [comment]
It isn't coming in via stdin of main process, so shouldn't be a problem.
Have you noticed an issue?

**Comment by turon** [comment]
@jwhui those extra checks fail as we expected.  Need to handle the
asynchronous prefixes changed event.

**Comment by turon** [comment]
@jwhui I marked those known failures XFAIL for now.

**Comment by jwhui** [comment]
@turon, we can merge if you are committed to resolving the XFAILs.

**Comment by turon** [comment]
@jwhui yes I'm signed up for that.  I just won't be able to start on it
until middle of next week.

**Comment by darconeous** [comment]
👍
